module/zfs/dbuf.c

```diff
@@ -797,9 +833,121 @@ dbuf_evict_one(void)
 		db->db_caching_status = DB_NO_CACHE;
 		dbuf_destroy(db);
 		DBUF_STAT_BUMP(cache_total_evicts);
-	} else {
-		multilist_sublist_unlock(mls);
+		evicted += size + usize;
+
+		mls = multilist_sublist_lock_idx(
+		    &dbuf_caches[DB_DBUF_CACHE].cache, idx);
+		db = multilist_sublist_prev(mls, marker);
 	}
+
+	multilist_sublist_remove(mls, marker);
+	multilist_sublist_unlock(mls);
+	kmem_cache_free(dbuf_kmem_cache, marker);
+}
+
+typedef struct evict_arg {
+	taskq_ent_t	tqe;
+	unsigned	idx;
+	uint64_t	bytes;
+} evict_arg_t;
+
+static void
+dbuf_evict_task(void *arg)
+{
+	evict_arg_t *eva = arg;
+	dbuf_evict_many(eva->bytes, eva->idx);
+}
+
+static void
+dbuf_evict(void)
+{
+	int64_t bytes = (zfs_refcount_count(&dbuf_caches[DB_DBUF_CACHE].size) -
+	    dbuf_cache_lowater_bytes());
+
+	if (bytes <= 0)
+		return;
+
+	unsigned idx = multilist_get_random_index(
+	    &dbuf_caches[DB_DBUF_CACHE].cache);
+
+	if (!dbuf_evict_parallel)
+		return (dbuf_evict_many(bytes, idx));
+
+	/*
+	 * Go to the parallel eviction.
+	 */
+	unsigned int num_sublists = multilist_get_num_sublists(
+	    &dbuf_caches[DB_DBUF_CACHE].cache);
+	evict_arg_t *evarg = kmem_zalloc(sizeof (*evarg) * num_sublists,
+	    KM_SLEEP);
+	/*
+	 *  How we scale
+	 *
+	 *  Example 1, # of chunks less than # of tasks.
+	 *  We have:
+	 *  - 4 tasks
+	 *  - 3 chunks
+	 *  - 3 full col
+	 *  - 0 low cols.
+	 *
+	 *  The first low col index is 3.
+	 *  The tasks #0-#2 evict 1 chunk each.
+	 *
+	 *    0 | 1 | 2 | 3 |
+	 *  +===+===+===+===+
+	 *  | x | x | x |   |
+	 *  +---+---+---+---+
+	 *
+	 *  Example 2, # of chunks more than # of tasks.
+	 *  We have:
+	 *  - 4 tasks
+	 *  - 9 chunks
+	 *  - 1 full col
+	 *  - 3 low cols
+	 *
+	 *  The first low col index is 1.
+	 *  The task #0 evicts 3 chunks, the others evict 2 chunks each.
+	 *
+	 *    0 | 1 | 2 | 3 |
+	 *  +===+===+===+===+
+	 *  | x | x | x | x |
+	 *  +---+---+---+---+
+	 *  | x | x | x | x |
+	 *  +---+---+---+---+
+	 *  | x |   |   |   |
+	 *  +---+---+---+---+
+	 */
+
+	/*
+	 * Compute number of tasks to run (n), first low col index (k),
+	 * normal and low bytes per task.
+	 */
+	uint64_t nchunks = ((bytes - 1) >> DBUF_MIN_EVICT_PERTASK_SHIFT) + 1;
+	unsigned n = nchunks < num_sublists ? nchunks : num_sublists;
+	uint64_t fullrows = nchunks / n;
+	unsigned lastrowcols = nchunks % n;
+	unsigned k = (lastrowcols ? lastrowcols : n);
+
+	uint64_t bytes_pertask_low = fullrows << DBUF_MIN_EVICT_PERTASK_SHIFT;
+	uint64_t bytes_pertask = bytes_pertask_low + (lastrowcols ?
+	    (1 << DBUF_MIN_EVICT_PERTASK_SHIFT) : 0);
```

Comment:
Same as for ARC seems over-engineered.  No need for tasks to be multiple of `1 << DBUF_MIN_EVICT_PERTASK_SHIFT`.